### DISCUSSION TITLE
Is it possible to create "snippets"?

### DISCUSSION THREAD

**Discussion by cglong** [chrisant996/clink#422] [Q&A] [answered]
I'm new to Clink, so apologies if this is a basic question. I'm wondering if Clink has support for something similar to [Visual Studio Code's user-defined snippets](https://code.visualstudio.com/docs/editor/userdefinedsnippets). As a basic example, I'd like to be able to write:

```console
C:\>home
```

then e.g. hit <kbd>Tab</kbd> and have my prompt automatically substitute this with:

```console
C:\>cd %USERPROFILE%
```

**Comment by chrisant996** [comment, marked as answer]
<kbd>Tab</kbd> is typically for completion.  What you're looking for is more like "replacement" not "completion".  For example, completion lets you start to type the first few characters of something, press <kbd>Tab</kbd>, and get help completing what you started typing.  E.g. `cd %userp`<kbd>Tab</kbd> would complete to `cd %USERPROFILE%`.

Also, the specific example `home`<kbd>Tab</kbd> -> `cd %USERPROFILE%` probably isn't a good example for "replacement snippets".  Because if you want to change to your user profile directory, you could just type `~`<kbd>Enter</kbd>.

A better fit in a command shell would be to create an alias.  Then you don't even have to manually expand it.

Create an alias with `doskey home=cd %userprofile%`.  If you want to expand to literally `cd %userprofile%` without also expanding the environment variable, then use an input file with a list of alias definitions, instead of defining it in the command line.  See `doskey /?` for more info, and use internet searches to learn more about `doskey`.

With the alias defined, then you can use `home`<kbd>Enter</kbd> to run the command, without even needing to manually expand it.

Or you can type `home`<kbd>Ctrl</kbd>-<kbd>Alt</kbd>-<kbd>F</kbd> to expand the alias (e.g. to `cd %userprofile%`) so that you can view it or edit it.

And you can also use it with completion:  `ho`<kbd>Tab</kbd> would try to complete `ho` based on files and doskey alias names.  It could expand to `home`, and then you could press <kbd>Enter</kbd> to run the alias, or expand it with <kbd>Ctrl</kbd>-<kbd>Alt</kbd>-<kbd>F</kbd>, or so on.

There are many alternatives as well, but that's the most common way to do what you asked about in a command shell.

**Reply by cglong** [reply]
Thank you for the detailed answer! This looks like exactly what I'm looking for 🙂 In particular, I can add parameters to macros created by `doskey` and pass those to the aliased command.
